cf-java-logging-support-servlet/src/main/java/com/sap/hcp/cf/logging/servlet/filter/AbstractLoggingFilter.java

```diff
@@ -0,0 +1,93 @@
+package com.sap.hcp.cf.logging.servlet.filter;
+
+import java.io.IOException;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public abstract class AbstractLoggingFilter implements Filter {
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException,
+                                                                                              ServletException {
+        if (request instanceof HttpServletRequest && response instanceof HttpServletResponse) {
+            doFilterRequest((HttpServletRequest) request, (HttpServletResponse) response, chain);
+        } else {
+            chain.doFilter(request, response);
+        }
+    }
+
+    /**
+     * Provides a default implementation for handling servlet requests already
+     * cast to {@link HttpServletRequest} and {@link HttpServletResponse}.
+     * Custom implementations of {@link AbstractLoggingFilter} should overwrite
+     * {@link #preProcess(HttpServletRequest, HttpServletResponse)} and/or
+     * {@link #postProcess(HttpServletRequest, HttpServletResponse)}.
+     * 
+     * @param request
+     *            cast as {@link HttpServletRequest}
+     * @param response
+     *            cast as {@link HttpServletResponse}
+     * @param chain
+     *            the {@link FilterChain} to continue request handling
+     * @throws IOException
+     * @throws ServletException
+     */
+    protected void doFilterRequest(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+                                                                                                                throws IOException,
+                                                                                                                ServletException {
+        try {
+            preProcess(request, response);
+            chain.doFilter(request, response);
+        } finally {
+            postProcess(request, response);
+
+        }
```

Comment:
```suggestion
        preProcess(request, response);
        try {
            chain.doFilter(request, response);
        } finally {
            postProcess(request, response);
        }
```
IMO a `postProcess` should only be called after a successful `preProcess`. At least that is the pattern in Java.

Reply:
The rational was, that `preProcess` is likely to change the MDC, i.e. "global" state. I wanted to ensure, that this is always cleaned. But your suggestion is a valid point. I will think about it.

Reply:
IMHO there are (at least) two _simple_ error handling models:
* A method cleans the created/changed intermediate state, when it throws an error. If recover is not possible, then an additional, externally controlled method typically can't do better.
* An object is regarded as corrupt. The all other methods should fail with illegal state or the object is not used any longer.

In your case then `preProcess` has either successfully set the MDC **or** clears it, if the implementation within `preProcess` has a failing further step.

Reply:
My problem is something like the AddHttpHeadersToLogContextFilter: https://github.com/SAP/cf-java-logging-support/blob/297072adaa9fb4bbc4b6a042440b60eeed9d93ef/cf-java-logging-support-servlet/src/main/java/com/sap/hcp/cf/logging/servlet/filter/AddHttpHeadersToLogContextFilter.java#L60-L68

```
   @Override
    protected void preProcess(HttpServletRequest request, HttpServletResponse response) {
        for (HttpHeader header: headers) {
            String headerValue = HttpHeaderUtilities.getHeaderValue(request, header);
            if (header.getField() != null && headerValue != null) {
                LogContext.add(header.getField(), headerValue);
            }
        }
    }
```

If `request.getHeader("...")` throws a RuntimeException already set MDC fields will not get reset and may leak state. The idea to have `preProcess` in the `try` block was to not have to implement a guard in all derived filters. Maybe renaming `preProcess` would be an option?

Reply:
The guard would be:
```Java
   @Override
    protected void preProcess(HttpServletRequest request, HttpServletResponse response) {
        try{
          for (HttpHeader header: headers) {
              String headerValue = HttpHeaderUtilities.getHeaderValue(request, header);
              if (header.getField() != null && headerValue != null) {
                  LogContext.add(header.getField(), headerValue);
              }
          }
        } catch( Exception e ) {
           postProcess( ... ); // or whatever is appropriate
           throw e;
        }
    }
```

I see your point. Then `postProcess` is semantically a `cleanup` to be called in any case. But with `cleanup` the naming would be asymmetric.

As I stated, the flow is unusual but for me your way of doing it is not a show-stopper.
Than it's a documentation thing to:
1. Not to do any cleanup in a failing `preProcess`
2. Even when `preProcess` failed with an exception `postProcess` will be called. If there is for example a counted resource usage (some streams for example don't allow a second close), then this state has to be handled by the cleanup implementation.

Reply:
How about renaming the functions to `onFilter` and `cleanUp` and add the documentation you suggested?

Reply:
Maybe `beforeFilter` instead of `onFilter`. In UI and event systems an `onXYZ(...)`  method is used to handle call completely. I would not change `cleanup` to camel case as it is a regular english word. 

(Side note: When nesting `javax.servlet.Filter` in your scenario instead of the `AbstractLoggingFilter` approach, then naturally the handling would be done in the `finally {}` block . Don't get me wrong - I don't intend to change the design, these are just my thoughts).

Reply:
Nesting `javax.servlet.Filter` would make them aware of the other filters,, if I understand your side note correctly. However, the architecture gives the knowledge about the ordering of filters to the filterChain, by using a visitor pattern. So nesting the filters would break that design.

Reply:
I implemented the renaming and imrpoved the documentation.